### PR TITLE
Travis CI/CD, PyPi release, tox + flake8 tests

### PR DESCRIPTION
**Pull request by leigh-johnson**
Hi!

First, thank you so much for your work on this gym! I'm ~5,000 episodes into my first training session using your [example ddqn agent](https://github.com/tawnkramer/donkey_gym/blob/master/examples/reinforcement_learning/ddqn.py). 

I write tutorials about TensorFlow + Raspberry Pi projects, and I'm currently working on a reinforcement learning example using Donkey Car and the OpenAI gym + Unity Sim you've created. To reduce the barrier to entry, I wanted to make it possible to `pip install gym-donkeycar`. 

Let me know if you're interested in integrating some of these changes.

### Summary

This PR contains some release tools, CI/CD pipeline, with an example wheel/tarball uploaded @ https://pypi.org/project/gym-donkeycar/. 

Releases can also be cut and uploaded from a local build:

```bash
bumpversion --new-version x.y.z 
make release
git push z.y.z
```

### CI/CD

https://travis-ci.org/leigh-johnson/gym-donkeycar

* Run tox + flake8 on all Pull Requests
* On new git tag creation, create dist and upload to PyPi

## Docs

https://gym-donkeycar.readthedocs.io/en/latest/

Local sphinx doc build:
```bash
make docs
```

## Breaking Changes

* Python package name `donkey_gym` => `gym_donkeycar` 
* PyPi package name `gym-donkeycar`
* 2.7-3.5 support missing (unsure if previously supported)

### CODE REVIEW

**Comment by tawnkramer** [comment]
Nice! You must be a devops guy? Appreciate all this work.

Does gym-donkeycar follow other gym like project naming conventions? I’m
fine either way just curious.

A pip install would be the next step. I just always do from github but I
appreciate. I’ve felt like the rl examples here needed some improvement.
I’ve had only moderate success with them. But not sure if that’s normal.
Let me know what you think.

I’ve felt like the binaries should come down automagically at some point.
Have any experience with that?

Cheers,
Tawn

**Comment by leigh-johnson** [comment]
Thank you! Release engineering is just one of the many hats I wear, haha.

The `gym-$package` naming convention comes from this document by OpenAI via [How to create new environments for Gym](https://github.com/openai/gym/blob/master/docs/creating-environments.md), and from what I've observed in other gym releases like `gym-duckietown` and `gym-sphero`.

The Unity sim binaries can be packaged too. Happy to help w/ that in a separate PR. A rough example would be...

```python
# setup.py
import platform
import setuptools

```
CUSTOM_COMMANDS = {
    'Linux':  [ ['wget', 'linux-foo.zip'] ],
    'Windows': [ ['wget', 'windows-foo.zip'] ],
    'Darwin':  [ ['curl', 'osx-foo.zip'] ],
}
class build(_build):
    sub_commands = _build.sub_commands + [('CustomCommands', None)]

class CustomCommands(setuptools.Command):
    def initialize_options(self):
        pass

    def finalize_options(self):
        pass

    def RunCustomCommand(self, command_list):
        p = subprocess.Popen(
            command_list,
            stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
        # Can use communicate(input='y\n'.encode()) if the command run requires
        # some confirmation.
        stdout_data, _ = p.communicate()
        print 'Command output: %s' % stdout_data
        if p.returncode != 0:
            raise RuntimeError('Command %s failed: exit code: %s' % (command_list, p.returncode))

    def run(self):
        system = platform.system()
        if system not in CUSTOM_COMMANDS.keys():
            raise NotImplemented('Unsupported platform {}. Supported platforms are Windows, Linux, OS X'.format(system))
        for command in CUSTOM_COMMANDS[system]:
            self.RunCustomCommand(command)

    cmdclass={
        # Command class instantiated and run during pip install scenarios.
        'build': build,
        'CustomCommands': CustomCommands,
    }
setup(
    ...
    keywords='gym_donkeycar',
    name='gym_donkeycar',
    ...
    cmdclass={
        'build': build,
        'CustomCommands': CustomCommands,
)
```

**Comment by leigh-johnson** [comment]
I think the RL examples you provided are very helpful, by the way! Thank you again for making them available. I'm interested in adding Tensorboard metrics in order to analyze further. 

For my own training sessions, my next steps are to implement DQN and DDQN [TF Agents](https://github.com/tensorflow/agents) in the gym-donkeycar environment. I implemented a DDQN to play Atari a few months ago, but this project is way more ambitious! Wish me luck 🤞Dumping some of [notes about q-learning techniques](https://drive.google.com/open?id=1feC7i6D3fbK0VGUPViVgKHBBn-aMT2LR), just in case that's of interest to you.

If you're curious what the TF-Agents API looks like, here's a [DQN reference implementation using CartPole](https://colab.research.google.com/github/tensorflow/agents/blob/master/tf_agents/colabs/1_dqn_tutorial.ipynb).
https://drive.google.com/open?id=1feC7i6D3fbK0VGUPViVgKHBBn-aMT2LR

Part of what I do as a [GDE in Machine Learning](https://developers.google.com/community/experts/) is help support the ecosystem around TensorFlow. That usually providing feedback on the TF ecosystem (TF-Agents), alpha/beta testing new offerings (TF-2.0), or by contributing to communities doing cool things (Donkey Car is incredible!!!)

**Comment by tawnkramer** [comment]
Tf agents. Nice. Need to take a long look there. Have not seen that.

I’ve deliberately made this project a bit more challenging. It’s always
real time and doesn’t pause for a learning session. Not sure that’s
standard. Pretty sure it’s not. But I like the real-time approach.

If need be we can add some synchronization.

Happy to see you are into Donkeycar. It’s a blast.

I will probably merge your PR when I get some time to look it over.

**Comment by tawnkramer** [comment]
Lots of nice work. I appreciate it. But I would not want to change the author and copyright to you. Can you please change that to my name?

**Comment by leigh-johnson** [comment]
Yikes, apologies for that! A lot of this diff is generated from a cookiecutter I use for pypi packages. Will fix. 

A summary of the commits coming in today:
* flake8 linting
* testing out the build/release system
* testing/fixing build installed via `pip install gym-donkeycar`

> Lots of nice work. I appreciate it. But I would not want to change the author and copyright to you. Can you please change that to my name?

I'd like to add you as a collaborator to the PyPi repo. Do you have an account / what's your username?

File: setup.py [hunk 1]
```diff
@@ -1,10 +1,54 @@
...
+    'pillow'
+]
+
+setup_requirements = ['pytest-runner', ]
+
+test_requirements = ['pytest', ]
+
+setup(
+    author='Tawn Kramer',
+    author_email='leigh@data-literate.com',
```
Comment (leigh-johnson) [inline setup.py]: @tawnkramer `author_email` is required for PyPi upload, so I left mine here. Let me know if you have an email you'd like to use. Note: this is public, so it should not be an email you use as a login/username.

Comment (tawnkramer) [reply setup.py]: tawnkramer@gmail.com is my email. tawnkramer is the user account. I have two-factor auth enabled here, so think we are ok.

**Pull request merged**